### ISSUE TITLE
Add support for `str` subclasses 

### ISSUE THREAD

**Issue by rpmcginty** [lidatong/dataclasses-json#348]
### Problem
as of v0.5.7, any subclassed `str` classes are instantiated using the default `str` class. 

#### Example
```python
class MyString(str):
    ...
    def my_useless_odd_number_detector(self) -> bool:
        return len(self) % 2 != 0

@dataclass_json
@dataclass
class MyDC:
    my_str: MyString

    def __post_init__(self):
        print(f"IS it odd? -> {self.my_str.my_useless_odd_number_detector()}")

# Error!
MyDC.from_dict(dict(my_str="123"))
```

To work around this, you must override the `decoder`. But this must be done in every dataclass that references that type. Alternatively, we can add this decoder to the global config. but this can be cumbersome if you are subclassing str many times. 

For context, I have a subclassed `str` into `ValidatedStr` which provides validation (regex, length constraints) against the string. I vend this `ValidatedStr` class for usage across other packages. I do not want to require others using this subclass to have to explicitly specify the decoder every time or update a global config.   

### Feature Request

Modify the `decode_dataclass` logic to further distinguish between custom subclassed `str` classes.

**Comment by george-zubrienko** [comment]
I think you should use composition instead of inheritance for this use case - would make life much easier. As for the library behaviour, I agree this code should not raise

**Comment by matt035343** [comment]
This seems to be fixed as of version 0.5.9. Please reopen if the issue persists.
